## MIIA-4203 MODELOS AVANZADOS PARA ANÁLISIS DE DATOS II


# Introducción a Sistemas de Recomendación

## Actividad 10


### Profesor: Camilo Franco (c.franco31@uniandes.edu.co)

## Actividad en grupos
### Nombres: Lupe Pizón, Camilo Angarita, Juan Andrés Cabal

**Instrucciones:** Por favor escriba los nombres de los integrantes de su grupo. Esta actividad debe ser entregada a más tardar dentro de 8 días, con la respuesta para los ejercicios y preguntas en cada numeral.

En este cuadernos estudiaremos una primera introducción a los sistemas de recomendación, examinando un modelo de recomendación en base a la popularidad, sin ningún grado de personalización. En futuras actividades exploraremos otros sistemas basados en contenido o en filtrado colaborativo, los cuales toman en cuenta las preferencias de los usuarios. 

Los sistemas de recomendación están entre las aplicaciones más populares de la ciencia de datos actualmente. El marco que ofrecen estos sistemas permite representar modelos de decisión junto con métodos estadístico/computacionales para **aprender y estimar** las preferencias de los usuarios. 


## Introducción

En esta introducción vamos a construir un modelo simple basado en popularidad. Aprenderemos nociones que nos permitirán construir modelos más complejos de recomendación. 

En este cuaderno vamos a trabajar con una base de datos de películas IMDB (https://www.imdb.com/) 

En términos generales, los sistemas de recomendación pueden clasificarse en 3 tipos:

- **Recomendadores simples:** genera recomendaciones para cada usuario según la popularidad del ítem o producto. Por ejemplo, proponer películas o un género en base a su popularidad o su crítica, donde una buena crítica aproxima una mayor probabilidad de que la película sea apreciada por una audiencia "promedio".
- **Recomendadores basados en contenido:** sugiere elementos similares en función de un elemento en particular, como por ejemplo el género, la sinopsis o el reparto (actores, etc). De esta manera, si el usuario tiene unas preferencias específicas sobre un ítem/artículo en particular, también podría tener preferencia por un ítem *similar*.
- **Recomendadores por filtrado colaborativo:** estos sistemas generan recomendaciones a partir de similitudes entre usuarios, prediciendo la calificación o preferencia de un usuario sobre un ítem, o un grupo de atributos de dicho ítem, según las calificaciones y preferencias observadas de otros usuarios. 

Por último se puede considerar también un cuarto tipo de recomendadores, que son los de tipo **Híbrido**. Estos hacen referencia al uso simultaneo de las distintas metodologias antes señaladas y de múltiples fuentes de información o tecnologías específicas, como pueden ser las redes sociales o fuentes de información geográfica, demográfica, etc.

Para profundizar en los sistemas de recomendación se pueden consultar las siguientes fuentes:
- Ch. Aggarwal. Recommender Systems. Springer, 2016.
- J. Leskovec, A. Rajaraman, J. Ullman. Mining of Massive Datasets. Cambridge University Press (http://www.mmds.org/)


## 1. Sistema de recomendación simple

Los recomendadores simples generan recomendaciones basadas en su popularidad, de acuerdo con una *preferencia promedio*. 

La pregunta es,

**¿cómo medimos la popularidad de un producto/ítem?**

Una opción es la de elegir una métrica determinada para puntuar las películas, y en base a esa métrica generar un valor de preferencia por el ítem. De esta manera es posible consturir un *orden lineal* o un *ranking estricto* con las mejores películas de la base de datos IMDB.

Pasos a seguir:
- Elegir la métrica con la que se va a calificar cada película. Este punto también se puede entender como un modelo para estimar las preferencias por una película.
- Calcular el valor de preferencia sobre cada película.
- Ordenar las películas según su valor o puntuación y obtener una lista con las mejores k peliculas.

Manos a la obra, primero carguemos el conjunto de datos como un data frame en la bilbioteca Pandas:

In [1]:
# Importamos la biblioteca Pandas
import pandas as pd

# Cargamos los datos de peliculas de la base de datos IMDB
metadata = pd.read_csv('movies_metadata.csv', low_memory=False)

print(metadata.shape)
      
list(metadata)

(45466, 24)


['adult',
 'belongs_to_collection',
 'budget',
 'genres',
 'homepage',
 'id',
 'imdb_id',
 'original_language',
 'original_title',
 'overview',
 'popularity',
 'poster_path',
 'production_companies',
 'production_countries',
 'release_date',
 'revenue',
 'runtime',
 'spoken_languages',
 'status',
 'tagline',
 'title',
 'video',
 'vote_average',
 'vote_count']

In [2]:
# Así se ven los datos
metadata.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   
2   1995-12-22          0.0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  
2          6.5       92.0  

[3 rows x 24 columns]

In [3]:
metadata[['original_title', 'popularity']][1:10]

original_title popularity
1                      Jumanji  17.015539
2             Grumpier Old Men    11.7129
3            Waiting to Exhale   3.859495
4  Father of the Bride Part II   8.387519
5                         Heat  17.924927
6                      Sabrina   6.677277
7                 Tom and Huck   2.561161
8                 Sudden Death    5.23158
9                    GoldenEye  14.686036

### Rating vs. Popularidad

Con esta información vamos a construir un sistema que recomiende películas basados directamente en su *rating*, o **voto promedio**. Sin embargo, es cierto que hay otro atributo relevante como lo es el de su **popularidad**. Por ejemplo, si un solo usuario califica con 10 una pelicula, pero 5000 usuarios le dan en promedio un 8.3, solo en base al rating la primera película sería considerada mejor que la segunda.


Bajo una perspectiva de *los grandes números*, la valoración o voto promedio de una película tiende a converger hacia un valor que refleje la calidad de la película a medida que el número de votos aumenta. Con poca información, o pocos votos, es más difícil estimar la calidad promedio de la película.  


Entonces, 

**¿qué tipo de métrica nos sirve para evaluar apropiadamente la preferencia por una película?** 

Tengamos en cuenta que en base a esta métrica se tendrán mejores o peores recomendaciones!




### Ejercicio 1.1

Examine los datos y proponga una métrica que permita estimar la preferencia promedio por una película.

In [24]:
metadata = metadata.drop(index=(metadata.loc[metadata['popularity'] == 'Beware Of Frost Bites'].index))
metadata['popularity'] = metadata['popularity'].astype(float)
metadata[['original_title', 'homepage', 'vote_count', 'vote_average', 'popularity']].describe()

vote_count  vote_average    popularity
count  45460.000000  45460.000000  45460.000000
mean     109.897338      5.618207      2.921478
std      491.310374      1.924216      6.005414
min        0.000000      0.000000      0.000000
25%        3.000000      5.000000      0.385948
50%       10.000000      6.000000      1.127685
75%       34.000000      6.800000      3.678902
max    14075.000000     10.000000    547.488298

Al examinar el número de votos, el promedio de votaciones y la popularidad de las películas es posible encontrar que, en promedio, el número de votos es igual a 109 mientras que el voto promedio es igual a 5.61 en una escala de 10. De igual forma, de las estadísticas descriptivas es posible observar que el promedio de popularidad de las películas es igual a 2.92, considerablemente un valor menor al rating promedio de las películas. 

Del análisis inicial es posible ver, de igual forma, que la variable popularidad, aunque puede servir como insumo para estimar la preferencia de una película, no se encuentra normalizada, siendo el valor máximo igual a 547.48 que corresponde a la película de los Minions, la cual tiene un rating promedio igual a 6.4. 

Ahora, dado que lo que se busca es proponer una métrica (diferente a la de rating_ponderado) que permita estimar la preferencia promedio por una película se hara uso de una calificación ponderada que tenga en cuenta tanto el ranting que recibe una película por su número de votos, así como el voto promedio global de las películas por el número de votos que recibe cada una de estas. Matemáticamente, se representa de la siguiente manera:

$$
\mu_i  = \left( \frac{v_i}{v_i+m_i} \right) R_i + \left(\frac{m_i}{v_i+m_i} \right) C 
$$

donde $v_i$ es el número de votos para la $i$-ésima película, $m_i$ es el mínimo de votos requerido para ser incluido en el cuadro, $R$ es el rating promedio de la pelicula y $C$ es el voto promedio global de todas las películas. 

In [26]:
def IMDB_rating(x, C):
    """
    Input:
    x: datos de rating y votacion de las peliculas
    m: minimo numero de votos
    C: promedio global
    Output:
    rating ponderado
    """
    m = x['vote_average'].max()
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

### Ponderación entre rating y popularidad

Una opción puede consistir en ponderar el rating que recibe una película por su número de votos. De esta manera se pueden prevenir situaciones donde una película con un voto de 10 obtenga una mejor calificación que una película con 1000 votos con un promedio de 9.5.  

De esta manera, el voto promedio ponderado $\mu_i$, de la $i$-ésima película lo definimos como:

$$
\mu_i  = \left( \frac{v_i}{v_{max}} \right) R_i 
$$

donde $v_i$ es el número de votos para la $i$-ésima película, $v_{max}$ es el máximo número de votos que recibe la película más popular, y $R$ es el rating promedio de la pelicula.


En la base de datos tenemos los valores de $v_i$ (vote_count) y $R$ (vote_average) para cada pelicula. Además podemos agregar un filtro para que considere las películas con un minimo número de votos $v_{min}$.


Nos quedamos con las columnas necesarias para calcular el rating ponderado más el link de la pagina (en caso de que esté disponible):

In [27]:
pelisR = metadata[['original_title', 'homepage', 'vote_count', 'vote_average']]
pelisR.shape

(45465, 4)

In [28]:
# Calificacion de peliculas basada en la ponderacion de su rating 
def rating_ponderado(x):
    """
    Input:
    x: datos de rating y votacion de las peliculas
    m: minimo numero de votos
    Output:
    rating ponderado
    """
    m = x['vote_average'].max()
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) 

Definimos una nueva respuesta 'score' y calculamos su valor con la funcion rating_ponderado().

A continuación ordenamos las películas basados en el score y ofrecemos el top 15 de películas preferidas:

In [29]:
import warnings
warnings.filterwarnings("ignore", category=Warning)

# si tiene al menos vmin votos, incluimos la pelicula
vmin = 1
pelisRF = pelisR.copy().loc[pelisR['vote_count'] >= vmin]
pelisRF['score'] = rating_ponderado(pelisRF)

# ordenamos las peliculas de acuerdo con el score
pelisRF = pelisRF.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisRF[['original_title', 'homepage', 'vote_count', 'vote_average', 'score']].head(15)

original_title  \
10309      Dilwale Dulhania Le Jayenge   
314           The Shawshank Redemption   
834                      The Godfather   
40251                            君の名は。   
39085                     Planet Earth   
12481                  The Dark Knight   
2843                        Fight Club   
292                       Pulp Fiction   
522                   Schindler's List   
23673                         Whiplash   
5481                          千と千尋の神隠し   
2211                   La vita è bella   
1178            The Godfather: Part II   
1152   One Flew Over the Cuckoo's Nest   
1176                            Psycho   

                                              homepage  vote_count  \
10309                                              NaN       661.0   
314                                                NaN      8358.0   
834                       http://www.thegodfather.com/      6024.0   
40251  https://www.funimationfilms.com/movie/yourname/      1030.0   
39085                                              NaN       176.0   
12481     http://thedarkknight.warnerbros.com/dvdsite/     12269.0   
2843        http://www.foxmovies.com/movies/fight-club      9678.0   
292                                                NaN      8670.0   
522                     http://www.schindlerslist.com/      4436.0   
23673                http://sonyclassics.com/whiplash/      4376.0   
5481            http://movies.disney.com/spirited-away      3968.0   
2211                                               NaN      3643.0   
1178                                               NaN      3418.0   
1152                                               NaN      3001.0   
1176                                               NaN      2405.0   

       vote_average     score  
10309           9.1  8.964382  
314             8.5  8.489842  
834             8.5  8.485913  
40251           8.5  8.418269  
39085           8.8  8.326882  
12481           8.3  8.293240  
2843            8.3  8.291433  
292             8.3  8.290438  
522             8.3  8.281332  
23673           8.3  8.281076  
5481            8.3  8.279135  
2211            8.3  8.277279  
1178            8.3  8.275788  
1152            8.3  8.272434  
1176            8.3  8.265631

### Examinemos el efecto de cambiar el mínimo numero de votos a considerar:

In [30]:
# si tiene al menos mil votos incluimos la pelicula
vmin = 1000
pelisRF = pelisR.copy().loc[pelisR['vote_count'] >= vmin]
pelisRF['score'] = rating_ponderado(pelisRF)

# ordenamos las peliculas de acuerdo con el score
pelisRF = pelisRF.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisRF[['original_title', 'homepage', 'vote_count', 'vote_average', 'score']].head(15)

original_title  \
314           The Shawshank Redemption   
834                      The Godfather   
40251                            君の名は。   
12481                  The Dark Knight   
2843                        Fight Club   
292                       Pulp Fiction   
522                   Schindler's List   
23673                         Whiplash   
5481                          千と千尋の神隠し   
2211                   La vita è bella   
1178            The Godfather: Part II   
1152   One Flew Over the Cuckoo's Nest   
1176                            Psycho   
1184       Once Upon a Time in America   
351                       Forrest Gump   

                                              homepage  vote_count  \
314                                                NaN      8358.0   
834                       http://www.thegodfather.com/      6024.0   
40251  https://www.funimationfilms.com/movie/yourname/      1030.0   
12481     http://thedarkknight.warnerbros.com/dvdsite/     12269.0   
2843        http://www.foxmovies.com/movies/fight-club      9678.0   
292                                                NaN      8670.0   
522                     http://www.schindlerslist.com/      4436.0   
23673                http://sonyclassics.com/whiplash/      4376.0   
5481            http://movies.disney.com/spirited-away      3968.0   
2211                                               NaN      3643.0   
1178                                               NaN      3418.0   
1152                                               NaN      3001.0   
1176                                               NaN      2405.0   
1184                                               NaN      1104.0   
351                                                NaN      8147.0   

       vote_average     score  
314             8.5  8.491364  
834             8.5  8.488023  
40251           8.5  8.430429  
12481           8.3  8.294254  
2843            8.3  8.292717  
292             8.3  8.291871  
522             8.3  8.284126  
23673           8.3  8.283909  
5481            8.3  8.282258  
2211            8.3  8.280679  
1178            8.3  8.279410  
1152            8.3  8.276558  
1176            8.3  8.270769  
1184            8.3  8.236584  
351             8.2  8.191454

Veamos en qué consiste esa pelicula en japonés:

In [31]:
import webbrowser

webbrowser.open(pelisRF['homepage'][40251]) 

True

### Ejercicio 1.2

Examine cómo varían los resultados si tomamos directamente la variable de popularidad.

Con el fin de hacer los resultados anteriores comparables, se examinará como varían los resultados si se toma directamente la variable de popularidad sobre aquellas películas que tengan al menos mil votos incluidos. En este caso, se observa un resultado diferente al que se obtiene al utilizar la métrica *ranting_ponderado*, siendo Minioms la primera película de la lista, seguido por Wonder Woman. De hecho, al examinar el top 15 de películas que se encuentran cuando se utiliza la variable popularidad, ninguna de estás hacen parte del top 15 que se encuentra cuando se utiliza el *ranting_ponderado*.

Este resultado muestran, en últimas, que la métrica que se utilice para evaluar apropiadamente la preferencia de una película puede mejorar o empeorar el resultado de la recomendación. En ese sentido, es importante entender la estructura de la base de datos y tener bajo consideración como se distribuyen los datos para construir un indicador apropiado que permita hacer una recomendación valida. 

In [34]:
pelisR = metadata[['original_title', 'homepage', 'vote_count', 'vote_average', 'popularity']]

# si tiene al menos mil votos incluimos la pelicula
vmin = 1000
pelisPF = pelisR.copy().loc[pelisR['vote_count'] >= vmin]

# ordenamos las peliculas de acuerdo con la popularidad
pelisPF = pelisPF.sort_values('popularity', ascending=False)

# Mostramos el top15 de las peliculas
pelisPF[['original_title', 'homepage', 'vote_count', 'vote_average', 'popularity']].head(15)

original_title  \
30700                                           Minions   
33356                                      Wonder Woman   
42222                              Beauty and the Beast   
43644                                       Baby Driver   
24455                                        Big Hero 6   
26564                                          Deadpool   
26566                    Guardians of the Galaxy Vol. 2   
14551                                            Avatar   
24351                                         John Wick   
23675                                         Gone Girl   
24873             The Hunger Games: Mockingjay - Part 1   
44274                    War for the Planet of the Apes   
26567                        Captain America: Civil War   
292                                        Pulp Fiction   
26560  Pirates of the Caribbean: Dead Men Tell No Tales   

                                                homepage  vote_count  \
30700                       http://www.minionsmovie.com/      4729.0   
33356             http://www.warnerbros.com/wonder-woman      5025.0   
42222  http://movies.disney.com/beauty-and-the-beast-...      5530.0   
43644                                                NaN      2083.0   
24455                http://movies.disney.com/big-hero-6      6289.0   
26564           http://www.foxmovies.com/movies/deadpool     11444.0   
26566  http://marvel.com/movies/movie/221/guardians_o...      4858.0   
14551                        http://www.avatarmovie.com/     12114.0   
24351                   http://www.johnwickthemovie.com/      5499.0   
23675                      http://www.gonegirlmovie.com/      6023.0   
24873                   http://www.thehungergames.movie/      5767.0   
44274  http://www.foxmovies.com/movies/war-for-the-pl...      1675.0   
26567           http://marvel.com/captainamericapremiere      7462.0   
292                                                  NaN      8670.0   
26560  http://pirates.disney.com/pirates-of-the-carib...      2814.0   

       vote_average  popularity  
30700           6.4  547.488298  
33356           7.2  294.337037  
42222           6.8  287.253654  
43644           7.2  228.032744  
24455           7.8  213.849907  
26564           7.4  187.860492  
26566           7.6  185.330992  
14551           7.2  185.070892  
24351           7.0  183.870374  
23675           7.9  154.801009  
24873           6.6  147.098006  
44274           6.7  146.161786  
26567           7.1  145.882135  
292             8.3  140.950236  
26560           6.6  133.827820

### Recomendación de peliculas mas populares por genero

In [41]:

metadata['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

0         [Animation, Comedy, Family]
1        [Adventure, Fantasy, Family]
2                   [Romance, Comedy]
3            [Comedy, Drama, Romance]
4                            [Comedy]
                     ...             
45461                 [Drama, Family]
45462                         [Drama]
45463       [Action, Drama, Thriller]
45464                              []
45465                              []
Name: genres, Length: 45465, dtype: object

In [43]:
import numpy as np
from ast import literal_eval

# trabajamos la informacion por generos
metadata['genres'] = metadata['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

# añadimos la variable del año
metadata['year'] = pd.to_datetime(metadata['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

metadata.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   

     imdb_id original_language    original_title  \
0  tt0114709                en         Toy Story   
1  tt0113497                en           Jumanji   
2  tt0113228                en  Grumpier Old Men   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  year  
0          7.7     5415.0  1995  
1          6.9     2413.0  1995  
2          6.5       92.0  1995  

[3 rows x 25 columns]

Primero nos quedamos con todos los generos:

In [46]:
metadata.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)

0        Animation
0           Comedy
0           Family
1        Adventure
1          Fantasy
           ...    
45461       Family
45462        Drama
45463       Action
45463        Drama
45463     Thriller
Length: 91102, dtype: object

In [47]:
generos = metadata.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
generos.name = 'genre'
gen_md = metadata.drop('genres', axis=1).join(generos)

gen_md.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   

                               homepage   id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story  862  tt0114709                en   
0  http://toystory.disney.com/toy-story  862  tt0114709                en   
0  http://toystory.disney.com/toy-story  862  tt0114709                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   

   popularity  ... runtime                          spoken_languages  \
0   21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]   
0   21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]   
0   21.946943  ...    81.0  [{'iso_639_1': 'en', 'name': 'English'}]   

     status tagline      title  video vote_average vote_count  year      genre  
0  Released     NaN  Toy Story  False          7.7     5415.0  1995  Animation  
0  Released     NaN  Toy Story  False          7.7     5415.0  1995     Comedy  
0  Released     NaN  Toy Story  False          7.7     5415.0  1995     Family  

[3 rows x 25 columns]

Construimos una funcion para un género particular y que tome en cuenta peliculas con un número vmin de votos:

In [48]:
def rec_gen(genero, vmin):
    df = gen_md[gen_md['genre'] == genero]
    v = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    R = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    m = df['vote_average'].max()
    
    pelisG = df[(df['vote_count'] >= vmin) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'overview', 'homepage']]
    pelisG['vote_count'] = pelisG['vote_count'].astype('int')
    pelisG['vote_average'] = pelisG['vote_average'].astype('int')
    
    pelisG['wr'] = v/m * R
    pelisG = pelisG.sort_values('wr', ascending=False).head(250)
    
    return pelisG

Veamos el Top-15 de recomendaciones en Ciencia Ficción:

In [49]:
k = 15
scifi = rec_gen('Science Fiction', 1000)
scifi.head(15)

title  year  vote_count  vote_average  \
15480                     Inception  2010       14075             8   
22879                  Interstellar  2014       11187             8   
14551                        Avatar  2009       12114             7   
17818                  The Avengers  2012       12000             7   
23753       Guardians of the Galaxy  2014       10014             7   
26553            Mad Max: Fury Road  2015        9629             7   
2458                     The Matrix  1999        9079             7   
12588                      Iron Man  2008        8951             7   
18244              The Hunger Games  2012        9634             6   
26555  Star Wars: The Force Awakens  2015        7993             7   
256                       Star Wars  1977        6778             8   
20830                    Iron Man 3  2013        8951             6   
25084                Jurassic World  2015        8842             6   
26567    Captain America: Civil War  2016        7462             7   
30051                   The Martian  2015        7442             7   

       popularity                                           overview  \
15480   29.108149  Cobb, a skilled thief who commits corporate es...   
22879   32.213481  Interstellar chronicles the adventures of a gr...   
14551  185.070892  In the 22nd century, a paraplegic Marine is di...   
17818   89.887648  When an unexpected enemy emerges and threatens...   
23753   53.291601  Light years from Earth, 26 years after being a...   
26553   29.361780  An apocalyptic story set in the furthest reach...   
2458    33.366332  Set in the 22nd century, The Matrix tells the ...   
12588   22.073099  After being held captive in an Afghan cave, bi...   
18244   20.031667  Every year in the ruins of what was once North...   
26555   31.626013  Thirty years after defeating the Galactic Empi...   
256     42.149697  Princess Leia is captured and held hostage by ...   
20830   23.721243  When Tony Stark's world is torn apart by a for...   
25084   32.790475  Twenty-two years after the events of Jurassic ...   
26567  145.882135  Following the events of Age of Ultron, the col...   
30051   25.629930  During a manned mission to Mars, Astronaut Mar...   

                                                homepage       wr  
15480              http://inceptionmovie.warnerbros.com/  11260.0  
22879                  http://www.interstellarmovie.net/   8949.6  
14551                        http://www.avatarmovie.com/   8479.8  
17818                  http://marvel.com/avengers_movie/   8400.0  
23753                        http://marvel.com/guardians   7009.8  
26553                        http://www.madmaxmovie.com/   6740.3  
2458                    http://www.warnerbros.com/matrix   6355.3  
12588                       http://www.ironmanmovie.com/   6265.7  
18244                   http://www.thehungergames.movie/   5780.4  
26555  http://www.starwars.com/films/star-wars-episod...   5595.1  
256    http://www.starwars.com/films/star-wars-episod...   5422.4  
20830                         http://marvel.com/ironman3   5370.6  
25084                      http://www.jurassicworld.com/   5305.2  
26567           http://marvel.com/captainamericapremiere   5223.4  
30051        http://www.foxmovies.com/movies/the-martian   5209.4

Inception e Interstellar aparecen en las dos primeras posiciones. Personalmente me gusta más Interstellar (si quieres ver un agujero negro, esta película es lo mejor que podrás conseguir), pero reconozco que Inception tiene mucho nivel. Podemos seguir refinando este tipo de recomendaciones prestando atención a los distintos atributos que tenemos disponibles sobre las películas. 


### Ejercicio 1.3

Elija el género de su interés y aplique la métrica propuesta en el Ejercicio 1.1

## Genero: Aventura

In [68]:
C = df['vote_average'].mean()
df = gen_md[gen_md['genre'] == 'Adventure']

# si tiene al menos mil votos incluimos la pelicula
pelisG1 = df[(df['vote_count'] >= vmin) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'overview', 'homepage', 'genre']]

# ordenamos las peliculas de acuerdo con la popularidad
pelisG1['score'] = IMDB_rating(pelisG1, C)
pelisG1 = pelisG1.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisG1.head(15)

title  year  vote_count  \
5481                                       Spirited Away  2001      3968.0   
1154                             The Empire Strikes Back  1980      5998.0   
9698                                Howl's Moving Castle  2004      2049.0   
2884                                   Princess Mononoke  1997      2041.0   
15480                                          Inception  2010     14075.0   
22879                                       Interstellar  2014     11187.0   
7000       The Lord of the Rings: The Return of the King  2003      8226.0   
256                                            Star Wars  1977      6778.0   
4863   The Lord of the Rings: The Fellowship of the Ring  2001      8892.0   
5814               The Lord of the Rings: The Two Towers  2002      7641.0   
1225                                  Back to the Future  1985      6239.0   
23753                            Guardians of the Galaxy  2014     10014.0   
17437       Harry Potter and the Deathly Hallows: Part 2  2011      6141.0   
3456                                           Gladiator  2000      5566.0   
1167                                  Return of the Jedi  1983      4763.0   

       vote_average  popularity  \
5481            8.3   41.048867   
1154            8.2   19.470959   
9698            8.2   16.136048   
2884            8.2   17.166725   
15480           8.1   29.108149   
22879           8.1   32.213481   
7000            8.1   29.324358   
256             8.1   42.149697   
4863            8.0   32.070725   
5814            8.0   29.423537   
1225            8.0   25.778509   
23753           7.9   53.291601   
17437           7.9   24.990737   
3456            7.9   23.186667   
1167            7.9   14.586087   

                                                overview  \
5481   A ten year old girl who wanders away from her ...   
1154   The epic saga continues as Luke Skywalker, in ...   
9698   When Sophie, a shy young woman, is cursed with...   
2884   Ashitaka, a prince of the disappearing Ainu tr...   
15480  Cobb, a skilled thief who commits corporate es...   
22879  Interstellar chronicles the adventures of a gr...   
7000   Aragorn is revealed as the heir to the ancient...   
256    Princess Leia is captured and held hostage by ...   
4863   Young hobbit Frodo Baggins, after inheriting a...   
5814   Frodo and Sam are trekking to Mordor to destro...   
1225   Eighties teenager Marty McFly is accidentally ...   
23753  Light years from Earth, 26 years after being a...   
17437  Harry, Ron and Hermione continue their quest t...   
3456   In the year 180, the death of emperor Marcus A...   
1167   As Rebel leaders map their strategy for an all...   

                                                homepage      genre     score  
5481              http://movies.disney.com/spirited-away  Adventure  8.294544  
1154   http://www.starwars.com/films/star-wars-episod...  Adventure  8.196526  
9698                                                 NaN  Adventure  8.189858  
2884                                                 NaN  Adventure  8.189818  
15480              http://inceptionmovie.warnerbros.com/  Adventure  8.098577  
22879                  http://www.interstellarmovie.net/  Adventure  8.098210  
7000                       http://www.lordoftherings.net  Adventure  8.097567  
256    http://www.starwars.com/films/star-wars-episod...  Adventure  8.097048  
4863                      http://www.lordoftherings.net/  Adventure  7.997842  
5814                      http://www.lordoftherings.net/  Adventure  7.997489  
1225   http://www.backtothefuture.com/movies/backtoth...  Adventure  7.996926  
23753                        http://marvel.com/guardians  Adventure  7.898167  
17437                         http://www.harrypotter.com  Adventure  7.897012  
3456                                                 NaN  Adventure  7.896704  
1167   http://www.starwars.com/films/star-wars-episod...  Adventure  7.896149

Spirited Away y The Empire Strikes Back aparecen en las dos primeras posiciones cuando se miran las películas cuyo genero es de aventura. Ambas, ciertamente, han sido reconocidas como grandes películas. 

## Genero: Familiar

In [75]:
C = df['vote_average'].mean()
df = gen_md[gen_md['genre'] == 'Family']

# si tiene al menos mil votos incluimos la pelicula
pelisG1 = df[(df['vote_count'] >= vmin) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'overview', 'homepage', 'genre']]

# ordenamos las peliculas de acuerdo con la popularidad
pelisG1['score'] = IMDB_rating(pelisG1, C)
pelisG1 = pelisG1.sort_values('score', ascending=False)

# Mostramos el top15 de las peliculas
pelisG1.head(15)

title  year  vote_count  \
5481                                  Spirited Away  2001      3968.0   
1225                             Back to the Future  1985      6239.0   
359                                   The Lion King  1994      5520.0   
5833                             My Neighbor Totoro  1988      1730.0   
926                           It's a Wonderful Life  1946      1103.0   
30315                                    Inside Out  2015      6737.0   
17437  Harry Potter and the Deathly Hallows: Part 2  2011      6141.0   
13724                                            Up  2009      7048.0   
12704                                        WALL·E  2008      6439.0   
24455                                    Big Hero 6  2014      6289.0   
7725       Harry Potter and the Prisoner of Azkaban  2004      6037.0   
0                                         Toy Story  1995      5415.0   
36253                                      Zootopia  2016      4961.0   
14625                           Hachi: A Dog's Tale  2009      1769.0   
6232                                   Finding Nemo  2003      6292.0   

       vote_average  popularity  \
5481            8.3   41.048867   
1225            8.0   25.778509   
359             8.0   21.605761   
5833            8.0   13.507299   
926             8.0   15.031588   
30315           7.9   23.985587   
17437           7.9   24.990737   
13724           7.8   19.330884   
12704           7.8   16.088366   
24455           7.8  213.849907   
7725            7.7   28.460276   
0               7.7   21.946943   
36253           7.7   26.024868   
14625           7.7    8.621359   
6232            7.6   25.497794   

                                                overview  \
5481   A ten year old girl who wanders away from her ...   
1225   Eighties teenager Marty McFly is accidentally ...   
359    A young lion cub named Simba can't wait to be ...   
5833   Two sisters move to the country with their fat...   
926    George Bailey has spent his entire life giving...   
30315  Growing up can be a bumpy road, and it's no ex...   
17437  Harry, Ron and Hermione continue their quest t...   
13724  Carl Fredricksen spent his entire life dreamin...   
12704  WALL·E is the last robot left on an Earth that...   
24455  The special bond that develops between plus-si...   
7725   Harry, Ron and Hermione return to Hogwarts for...   
0      Led by Woody, Andy's toys live happily in his ...   
36253  Determined to prove herself, Officer Judy Hopp...   
14625  A drama based on the true story of a college p...   
6232   Nemo, an adventurous young clownfish, is unexp...   

                                                homepage   genre     score  
5481              http://movies.disney.com/spirited-away  Family  8.294684  
1225   http://www.backtothefuture.com/movies/backtoth...  Family  7.997015  
359               http://movies.disney.com/the-lion-king  Family  7.996627  
5833                                                 NaN  Family  7.989273  
926                                                  NaN  Family  7.983221  
30315                http://movies.disney.com/inside-out  Family  7.897359  
17437                         http://www.harrypotter.com  Family  7.897103  
13724            http://disney.go.com/disneypictures/up/  Family  7.797593  
12704        http://disney.go.com/disneypictures/wall-e/  Family  7.797365  
24455                http://movies.disney.com/big-hero-6  Family  7.797303  
7725                                                 NaN  Family  7.697327  
0                   http://toystory.disney.com/toy-story  Family  7.697021  
36253                  http://movies.disney.com/zootopia  Family  7.696749  
14625                                                NaN  Family  7.690909  
6232               http://movies.disney.com/finding-nemo  Family  7.597567

Al seleccionar un segundo genero: "Familiar" Spirited Away y The Empire Strikes Back aparecen nuevamente en las dos primeras posiciones. Lo que refuerza la popularidad de las dos peliculas